protocols/kad/src/handler.rs

```diff
@@ -468,16 +476,88 @@ pub struct KademliaRequestId {
 struct UniqueConnecId(u64);
 
 impl KademliaHandler {
-    /// Create a [`KademliaHandler`] using the given configuration.
-    pub fn new(
-        config: KademliaHandlerConfig,
+    /// Create a [`KademliaHandler`] for a new inbound connection.
+    pub fn new_inbound(
+        protocol_config: KademliaProtocolConfig,
+        idle_timeout: Duration,
+        local_addr: Multiaddr,
+        remote_addr: Multiaddr,
+        remote_peer_id: PeerId,
+        external_addresses: Vec<&Multiaddr>,
+    ) -> Self {
+        let mode = match external_addresses.as_slice() {
+            [] => {
+                log::debug!("Operating in client-mode on new inbound connection from peer {remote_peer_id} because we don't have any confirmed external addresses");
+
+                Mode::Client
+            }
+            [confirmed_external_addr] => {
+                // This may not be true if we are listening on multiple interfaces but it is a good enough approximation for now.
+                log::debug!("Operating in server-mode on new inbound connection from peer {remote_peer_id} under the assumption that {confirmed_external_addr} routes to {local_addr}");
+
+                Mode::Server
+            }
+            confirmed_external_addresses => {
+                let confirmed_external_addresses = confirmed_external_addresses
+                    .iter()
+                    .map(|addr| addr.to_string())
+                    .collect::<Vec<_>>()
+                    .join(", ");
+
+                log::debug!("Operating in server-mode on new inbound connection from peer {remote_peer_id} under the assumption that one of [{confirmed_external_addresses}] routes to {local_addr}");
+
+                Mode::Server
+            }
```

Comment:
This is where we are making assumptions. A few things we could do is:

- Compare the protocols of the local and remote address. If our confirmed external address is TCP and the remote dialed us via QUIC, then we shouldn't advertise kademlia on this connection.
- More generally, it would be good to track a mapping of external addresses and which local interface they belong to, then we could accurately decide here.

Reply:
In my eyes, for the sake of simplicity, our first iteration should have the same `Mode` across all connections (as implemented above). I don't feel comfortable with the suggested matching of local and external address, especially when it comes to compatibility across versions and implementations.

As a downside, this might lead to the local node being added to remote routing tables even though the remote node can not connect to the local node. The remote node will eventually notice and remove the local node from its routing table. Kademlia's routing table mechanism of keeping the oldest online nodes helps us here.

Reply:
This version has the same `Mode` across all connections. I send a `ReconfigureMode` message to all connections. This matching here only differs in the log message. We don't inspect the addresses themselves at all. The logic essentially is:

- Do we have at least 1 external address? Switch into server mode.
- Do we have no external addresses? Switch into client mode.

Reply:
I think we are on the same page here, there just seems to be a miscommunication.

> A few things we could do is:

I am suggesting to not do the initial suggestions, but to keep the logic as is. The only change I would like to see is that we treat outbound connections just like inbound connections. More specifically, when changing to `Mode::Server`, that we do so on outbound connections as well. See https://github.com/libp2p/rust-libp2p/pull/3877/files#r1204946875.